ejb3/src/main/java/org/jboss/as/ejb3/timerservice/TimerServiceImpl.java

```diff
@@ -1180,6 +1212,54 @@ public void afterCompletion(final int status) {
         }
     }
 
+    private class TaskPostPersist extends java.util.TimerTask {
+        private final TimerImpl timer;
+        private int triesCounter = 0;
+
+        TaskPostPersist(TimerImpl timer) {
+            this.timer = timer;
+        }
+
+        TaskPostPersist(TimerImpl timer, int triesCounter) {
+            this.timer = timer;
+            this.triesCounter = triesCounter;
+        }
+
+        @Override
+        public void run() {
+            final ExecutorService executor = executorServiceInjectedValue.getOptionalValue();
+            if (executor != null) {
+                executor.submit(this::persistTimer);
+            }
+        }
+
+        void persistTimer() {
+            final ContextTransactionManager transactionManager = ContextTransactionManager.getInstance();
+            try {
+                transactionManager.begin();
+                timerPersistence.getValue().persistTimer(timer);
+                transactionManager.commit();
+            } catch (Exception e) {
+                try {
+                    transactionManager.rollback();
+                } catch (Exception ee) {
+                    // omit;
+                }
+                EJB3_TIMER_LOGGER.exceptionRunningTimerTask(timer, timer.getTimedObjectId(), e);
+                Date nextExpiration = timer.getNextExpiration();
+                if( nextExpiration != null ){
+                    if( triesCounter++ < MAX_RETRY ) {
+                        long nextTryDelay = nextExpiration.getTime() - System.currentTimeMillis() / 2;
```

Comment:
I still don't think this heuristic makes sense. I think it should be using a randomised back off approach, so the first attempt should happen very quickly after the fail, then each subsequent attempt takes longer. I think something like:

- First retry is immediate
- Subsequent are C * N ms, where C is some specific constant (maybe 1s by default, but it should be configurable), and N is the attempt number

Reply:
OK, I agree. I've made changes to simplify algorithm.

First - all repetitions better to finish before the next timer expiration.
Second - simply divide the whole elapsed time till the next expiration (at the time of first error) to the number of repetitions and use it every time. 
Third - log an error message even for the single - action timer

Please see my last commit